front/lib/resources/membership_resource.ts

```diff
@@ -0,0 +1,344 @@
+}> & {
+  roles?: MembershipRoleType[];
+};
+
+// Attributes are marked as read-only to reflect the stateless nature of our Resource.
+// This design will be moved up to BaseResource once we transition away from Sequelize.
+// eslint-disable-next-line @typescript-eslint/no-empty-interface
+export interface MembershipResource
+  extends ReadonlyAttributesType<MembershipModel> {}
+export class MembershipResource extends BaseResource<MembershipModel> {
+  static model: ModelStatic<MembershipModel> = MembershipModel;
+
+  constructor(
+    model: ModelStatic<MembershipModel>,
+    blob: Attributes<MembershipModel>
+  ) {
+    super(MembershipModel, blob);
+  }
+
+  static async getActiveMemberships({
+    userIds,
+    workspace,
+    roles,
+  }: GetMembershipsOptions): Promise<MembershipResource[]> {
+    const whereClause: WhereOptions<InferAttributes<MembershipModel>> = {
+      startAt: {
+        [Op.lte]: new Date(),
+      },
+      endAt: {
+        [Op.or]: [{ [Op.eq]: null }, { [Op.gte]: new Date() }],
+      },
+    };
+
+    if (userIds) {
+      whereClause.userId = userIds;
+    }
+    if (workspace) {
+      whereClause.workspaceId = workspace.id;
+    }
+    if (roles) {
+      whereClause.role = {
+        [Op.in]: roles,
+      };
+    }
+
+    const memberships = await MembershipModel.findAll({
+      where: whereClause,
+    });
+
+    return memberships.map(
+      (membership) => new MembershipResource(MembershipModel, membership.get())
+    );
+  }
+
+  static async getLatestMemberships({
+    userIds,
+    workspace,
+    roles,
+  }: GetMembershipsOptions): Promise<MembershipResource[]> {
+    const orderedResourcesFromModels = (resources: MembershipModel[]) =>
+      resources
+        .sort((a, b) => a.startAt.getTime() - b.startAt.getTime())
+        .map(
+          (resource) => new MembershipResource(MembershipModel, resource.get())
+        );
+
+    const where: WhereOptions<InferAttributes<MembershipModel>> = {};
+    if (roles) {
+      where.role = {
+        [Op.in]: roles,
+      };
+    }
+    if (userIds) {
+      where.userId = {
+        [Op.in]: userIds,
+      };
+    }
+    if (workspace) {
+      where.workspaceId = workspace.id;
+    }
+
+    if (!workspace && !userIds?.length) {
+      throw new Error("At least one of workspace or userIds must be provided.");
+    }
+    if (userIds && !userIds.length) return [];
+
+    // Get all the memberships matching the criteria.
+    const memberships = await MembershipModel.findAll({
+      where,
+      order: [["startAt", "DESC"]],
+    });
+    // Then, we only keep the latest membership for each (user, workspace).
+    const latestMembershipByUserAndWorkspace = new Map<
+      string,
+      MembershipModel
+    >();
+    for (const m of memberships) {
+      const key = `${m.userId}__${m.workspaceId}`;
+      const latest = latestMembershipByUserAndWorkspace.get(key);
+      if (!latest || latest.startAt < m.startAt) {
```

Comment:
Technically equivalent, since you cannot create a new membership if there is already an overlapping one. If you have an active membership (i.e non-ended), it is necessarily the one with the greatest startAt, so it simplifies the code to only check for startAt.